libafl/src/stages/mod.rs

```diff
@@ -541,38 +548,128 @@ pub mod pybind {
-    fn progress_mut(state: &mut S) -> Result<&mut Self, Error>;
+    fn progress_mut<'a>(state: &'a mut S, stage: &ST) -> Result<&'a mut Self, Error>;
 }
 
-impl<S> StageProgress<S> for () {
-    fn initialize_progress(_state: &mut S) -> Result<(), Error> {
+impl<S, ST> StageProgress<S, ST> for () {
+    fn initialize_progress(_state: &mut S, _stage: &ST) -> Result<(), Error> {
         Ok(())
     }
 
-    fn clear_progress(_state: &mut S) -> Result<(), Error> {
+    fn clear_progress(_state: &mut S, _stage: &ST) -> Result<(), Error> {
         Ok(())
     }
 
-    fn progress(_state: &S) -> Result<&Self, Error> {
+    fn progress<'a>(_state: &'a S, _stage: &ST) -> Result<&'a Self, Error> {
         unimplemented!("The empty tuple resumable stage status should never be queried")
     }
 
-    fn progress_mut(_state: &mut S) -> Result<&mut Self, Error> {
+    fn progress_mut<'a>(_state: &'a mut S, _stage: &ST) -> Result<&'a mut Self, Error> {
         unimplemented!("The empty tuple resumable stage status should never be queried")
     }
 }
 
+/// Progress which permits a fixed amount of resumes per round of fuzzing. If this amount is ever
+/// exceeded, the input will no longer be executed by this stage.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+pub struct LimitedTriesProgress {
+    tries_remaining: Option<usize>,
+    skipped: HashSet<CorpusId>,
+}
+
+impl_serdeany!(LimitedTriesProgress);
+
+/// Stage which specifies a certain amount of tries over which a scheduled input will be used. To
+/// be used in combination with [`LimitedTriesProgress`].
+pub trait RetryingStage {
+    /// The number of tries permitted for each testcase.
+    fn initial_tries(&self) -> usize;
+}
+
+impl<S, ST> StageProgress<S, ST> for LimitedTriesProgress
+where
+    S: HasNamedMetadata,
+    ST: RetryingStage,
+{
+    fn initialize_progress(state: &mut S, stage: &ST) -> Result<(), Error> {
+        if let Ok(metadata) = state.named_metadata_mut::<Self>(core::any::type_name_of_val(stage)) {
+            if let Some(ref mut remaining) = metadata.tries_remaining {
+                *remaining = remaining.checked_sub(1).ok_or_else(|| {
+                    Error::illegal_state(
+                        "Attempted further retries after we had already gotten to none remaining.",
+                    )
+                })?;
+            } else {
+                metadata.tries_remaining = Some(stage.initial_tries());
+            }
+        } else {
+            state.add_named_metadata(
+                Self {
+                    tries_remaining: Some(stage.initial_tries()),
+                    skipped: HashSet::new(),
+                },
+                core::any::type_name_of_val(stage),
+            );
+        }
+        Ok(())
+    }
+
+    fn clear_progress(state: &mut S, stage: &ST) -> Result<(), Error> {
+        let metadata = state.named_metadata_mut::<Self>(core::any::type_name_of_val(stage))?;
```

Comment:
I think we should look up the the metadata once in the stage, every hashmap lookup is pretty expensive

Reply:
Or better yet, cache it locally in the stage instance? No need to hit the map every round if we didn't restart (`Rc<RefCell<>>` that guy into an Option or something)

Reply:
If we store this anywhere but in the state, it won't be recovered on resume, so we must eat the lookup. However, given the relative cost of tracing and that this happens once per stage initialisation (either the first time the stage is called for this fuzz cycle or during resume), I don't think this is going to be a perf issue.

Reply:
No we can look it up once if it's `None` (so after restart), else we just reuse the pre-looked-up one

Reply:
But this doesn't change how many lookups we do, no? We end up requesting the progress the same amount of times in the end.

Reply:
but ok maybe not necessary